mirrord-layer/tests/integration.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    collections::HashMap,
+    process::Stdio,
+    time::{Duration, Instant},
+};
+
+use actix_codec::Framed;
+use futures::{stream::StreamExt, SinkExt};
+use mirrord_protocol::{
+    tcp::{DaemonTcp, LayerTcp, NewTcpConnection, TcpData},
+    ClientMessage, DaemonCodec, DaemonMessage,
+};
+use rstest::rstest;
+use tokio::{
+    io::AsyncWriteExt,
+    net::{TcpListener, TcpStream},
+    process::Command,
+};
+
+/// Run python with shell resolving to find the actual executable.
+///
+/// This is to help tests that run python with mirrord work locally on systems with pyenv.
+/// If we run `python3` on a system with pyenv the first executed is not python but bash. On mac
+/// that prevents the layer from loading because of SIP.
+async fn get_python3_executable() -> String {
+    let mut python = Command::new("python3")
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()
+        .unwrap();
+    let child_stdin = python.stdin.as_mut().unwrap();
+    child_stdin
+        .write_all(b"import sys\nprint(sys.executable)")
+        .await
+        .unwrap();
+    let output = python.wait_with_output().await.unwrap();
+    String::from_utf8_lossy(&output.stdout).to_string()
+}
+
+struct LayerConnection {
+    codec: Framed<TcpStream, DaemonCodec>,
+    num_connections: u64,
+}
+
+impl LayerConnection {
+    /// Accept a connection from the libraries and verifies the first message it is supposed to send
+    /// to the agent - GetEnvVarsRequest. Send back a response.
+    /// Return the codec of the accepted stream.
+    async fn accept_library_connection(listener: &TcpListener) -> Framed<TcpStream, DaemonCodec> {
+        let (stream, _) = listener.accept().await.unwrap();
+        let mut codec = Framed::new(stream, DaemonCodec::new());
+        let msg = codec.next().await.unwrap().unwrap();
+        if let ClientMessage::GetEnvVarsRequest(request) = msg {
+            assert!(request.env_vars_filter.is_empty());
+            assert_eq!(request.env_vars_select.len(), 1);
+            assert!(request.env_vars_select.contains("*"));
+        } else {
+            panic!("unexpected request {:?}", msg)
+        }
+        codec
+            .send(DaemonMessage::GetEnvVarsResponse(Ok(HashMap::new())))
+            .await
+            .unwrap();
+        codec
+    }
+
+    /// Accept the library's connection and verify initial ENV message.
+    /// Handle python's 2 process behaviour.
+    async fn get_initialized_connection(listener: &TcpListener) -> LayerConnection {
+        let mut codec = Self::accept_library_connection(listener).await;
+        let msg = match codec.next().await {
+            Some(option) => option.unwrap(),
```

Comment:
I'd assert this behavior instead of making it a fallback (i.e expect it to happen or expect it not to happen based on current run parameters.)

Reply:
I made the test require two connections when running with the python server and [it looks like there is no second connection on the CI](https://github.com/metalbear-co/mirrord/actions/runs/3188716994/jobs/5201728879#step:8:35). Can it be that the python on the CI does not fork?
The test does pass locally.

Reply:
On the CI there is only one connection also on mac. I'm not sure on what cases there are two connections and on what  cases there's only one. Should we maybe return to the fallback strategy?

Reply:
weird, okay then :octocat: